### DISCUSSION TITLE
How to import 3rd party css dynamically

### DISCUSSION THREAD

**Discussion by dpatel4-lat** [vercel/next.js#17068] [Help] [answered]
What is the best way to load 3rd party CSS dynamically? This is a follow-up on next@^9.5.4-canary.10
If I do something like below then I get this error

```
import dynamic from 'next/dynamic'
import { Slider } from "@reach/slider";

const DynamicComponent = dynamic(() => import('@reach/slider/styles.css'))

function Home() {
  return (
    <div>
      <Header />
      <DynamicComponent />
      <p>HOME PAGE is here!</p>
    </div>
  )
}

export default Home
```

`Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.`

Any help/suggestions on webpack with css-loader would be great.

**Comment by jamesmosier** [comment]
`next/dynamic` can only be used for React components. 

Next works with the regular `import` spec ([recently merged PR](https://github.com/vercel/next.js/pull/16803/files) with the updated docs, should land on nextjs.org/docs soon-ish I'm sure).

But I don't believe `import` will work with a CSS file, although I could be wrong. Next will bundle these CSS imports into separate CSS files. Is there any reason you want to dynamically import a CSS file? 

According to this [PR that is only on canary](https://github.com/vercel/next.js/pull/16993) that landed this update, you should just be able to import your package's CSS file and it'll "just work".

**Reply by techandmedia** [reply]
Hi, sorry to bump this old thread. Dynamically import a css file is usefull when you have a SAAS app with multiple domains on a single app

So how can I do this?
`
const hostname = window.location.hostname;
if(hostname.includes('tenant-a'){
  import tenant-a.css
} else if(hostname.includes('tenant-b'){
  import tenant-b.css
} else {
  import generic.css
}
`

**Comment by dpatel4-lat** [comment]
We have generic React component library that is used on multiple applications (Next and non-Next). Each component has 2 CSS files based on theme. We use dynamic routing and import those components in the pages based on the theme. This [PR that is only on canary](https://github.com/vercel/next.js/pull/16993) solves part of the problem. We can import CSS based on theme that is needed for the page. But it still leaves the unused CSS. 

```
├── pages
 │   ├── [slug-theme1].js
 │    |           
 │   └── [slug-theme2].js

```

slug-theme1.js is dynamic route. It imports all components dynamically but I don't know how to include its css dynamically. I don't want to import all components css in that page. Do we need to create a corresponding component in Next for each 3rd party component to solve this issue?

**Comment by dpatel4-lat** [comment, marked as answer]
We solved it with latest Canary release of NextJS + in components side we bundled the JS with CSS via webpack. 

This is because we control those 3rd party components outside Next.

**Reply by rehanpunjwani** [reply]
@dpatel4-lat i want to import global css files based on the language of the app. Is there any way to import css dynamically?

**Reply by nitinjainhcl** [reply]
@dpatel4-lat  would you like to share your solution about how you bundled JS with CSS for your components, which later can be dynamically loaded into NextJS project

**Comment by cristian-eriomenco** [comment]
Can you delete this question or at least provide an actual answer to it?

**Reply by garyhe0780** [reply]
If you are using `app dir` mode, it's easy to put in your `layout.tsx` file, something like this:
```javascript
export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode
}>) {
  return (
    <html lang="en"> 
      <Fragment>
        <head>
          <link href="{your css file link}" rel='stylesheet' />
        </head>
      </Fragment>
      <body>
        {children}
      </body>
    </html>
  )
}
```

I want to do this is because my css file is a list of themes variables which can modify by users, so I can't baked into my source file.
